lib/pure/terminal.nim

```diff
@@ -727,6 +727,12 @@ proc getch*(): char =
       doAssert(readConsoleInput(fd, addr(keyEvent), 1, addr(numRead)) != 0)
       if numRead == 0 or keyEvent.eventType != 1 or keyEvent.bKeyDown == 0:
         continue
+      if keyEvent.uChar == 0:
+        # ignore the first SHIFT_PRESSED event - the subsequent read
+        # has the result. Details:
+        # https://github.com/nim-lang/Nim/issues/7764
+        if (keyEvent.dwControlKeyState and SHIFT_PRESSED) == SHIFT_PRESSED:
```

Comment:
What about Ctrl, Alt, Caps Lock, Num Lock, Scroll Lock?
Probably, they too should be filtered.

Reply:
The original motivation behind changing getch() was because `readPasswordFromStdIn` uses a nim implementation of getch() on windows which would produce non-printable characters when a key modifier like SHIFT is pressed since `readPasswordFromStdIn` didn't account for modifier keys beyond delete and carriage return.

Thinking about it further I think `getch()` changes should be reverted and `readPasswordFromStdIn` changed to account for modifier keys. 

This allows users of nim `getch()` on windows to look out for `getch()` returning 0 as a signal to check for modifiers as in the quick and dirty example below:

```nim
import terminal

when isMainModule and defined(windows):
  const 
    VK_LSHIFT = 160
    VK_LCONTROL = 162
    VK_RSHIFT = 161
    VK_RCONTROL = 163
    VK_LEFT = 37
    VK_UP = 38
    VK_RIGHT = 39
    VK_DOWN = 40    
  
  proc getKeyState*(nVirtKey: int32): int16 {.stdcall, dynlib: "user32", importc: "GetKeyState".}
  while true:
    let c = getch();
    if c == 'Q':
      break;
    elif ord(c) == 0:
      if getKeyState(VK_LEFT) < 0: echo "ARROW LEFT"
      if getKeyState(VK_RIGHT) < 0: echo "ARROW RIGHT"
      if getKeyState(VK_UP) < 0: echo "ARROW UP"
      if getKeyState(VK_DOWN) < 0: echo "ARROW DOWN"
      if getKeyState(VK_LSHIFT) < 0: echo "SHIFT LEFT"
      if getKeyState(VK_RSHIFT) < 0: echo "SHIFT RIGHT"
      if getKeyState(VK_LCONTROL) < 0: echo "CONTROL LEFT"
      if getKeyState(VK_RCONTROL) < 0: echo "CONTROL RIGHT"
      # and so on..
    else:
      echo ord(c), "'", c, "'" 
```

Reply:
Agreed but the point is if someone wants or needs to they can :) The issue at hand is should getch return everything which `readConsoleInput()` returns. After spending some time with the issue I believe it's better to return everything and leave the users to handle the input as they wish.

Following that logic `readPasswordFromStdin` should account for getch() returning 0x0 in uChar by ignoring it. Agreed?

Reply:
``getch``'s name implies that it only returns characters. IMO it should only do that and ignore everything else.

Reply:
That's legitimate too.  I based the current proposal on #6966 where Araq mentions:
 > I think this needs to be reverted. A better getch that can also return control keys should be created. A standard lib for keyboard keys would also make interop with other UI libraries easier.

Returning control keys directly is not possible when returning a char (because the information is just not there) but returning the modifier in uChar does allow for retrieving the modifier key information. We could also modify getch() to return extended information in a structure or allow for an optional callback procedure to be passed as a parameter.  What's clear is the current behaviour is not adequate. Returning everything would need to be documented. 

Any further opinions?

Reply:
Happy for ``getch`` to return extended information but it should be renamed then I think.

Reply:
Well possibly we should stick to the principle of least surprise and mimic what getch() / _getch() would return on windows which is:

SHIFT+ regular Key yields the key value without modifier information - (SHIFT+'a') returns 'A'
EXTENDED+Key yields a modifier code first and then the key value in the next call to get() - (pressing CTRL+Right Arrow)  0xE0 and then 116. pressing Right Arrow yields 0xE0 and then 77, and so on.

Function keys like F1,F2, and so on produce first a key modifier of 0x0 and then a key code. F1 return 0x0, and on the next call 59, F2 0x0 and on the next call 60, and so on.

SHIFT+key,ALT+key,CTRL+key and so on return the key code without modifier information unless an extended key is pressed after SHIFT, ALT, CTRL and so on.